DOC: Add missing imports, definitions and dummy file

pdebuyl opened the pull request:
Related to #14970 

Here, I fix the rst tests for `distutils.rst`:
- Add dummy file foo.c at runtime
- I added missing imports in an explicit manner, so as not to hide the information for readers

eric-wieser commented on doc/source/reference/distutils.rst:
```diff
@@ -117,6 +120,8 @@ install the C library, you just use the method `add_installed_library` instead o
 ``install_dir`` argument::
 
+  >>> import numpy.distutils.misc_util
+  >>> config = np.distutils.misc_util.Configuration(None, '', '.')
```
I think we either need to adjust the text to explain these lines, or hide them in a comment like the one at the top

pdebuyl replied:
I hesitated between the two. I can move both lines to the hidden block at the top.

eric-wieser commented on doc/source/reference/distutils.rst:
```diff
@@ -1,5 +1,7 @@
 .. for doctests
    >>> with open('foo.c', 'w') as f: pass
+   >>> import numpy.distutils.misc_util
+   >>> config = np.distutils.misc_util.Configuration(None, '', '.')
```
Can you move this block down to right above the example that needs it? I don't think putting this at the top adds clarity.

pdebuyl commented:
Should I squash the commits?

eric-wieser commented:
Sure, saves me doing it.

Can you add a `DOC:` prefix to the commit message, and make clear in the message that the goal is to make the rst files pass refguide?

pdebuyl committed:
DOC: fix imports and defs to pass refguide checks

Add empty file foo.c at runtime to run distutils config.

Add import of numpy.distutils.misc_util.

Add np.distutils.misc_util prefix where needed.

Fix syntax for ** argument.

charris commented on doc/source/reference/distutils.rst:
```diff
@@ -117,6 +120,11 @@ install the C library, you just use the method `add_installed_library` instead o
 ``install_dir`` argument::
 
+.. for doctests
+
```
I think this would read easier without the blank lines.

pdebuyl replied:
sphinx complains about the version without a blank line, see failed circle ci

charris replied:
Interesting. I based the suggestion on other uses of `.. for doctests` in the documentation and it seems they may be in error.

charris replied:
Actually, looks like the indentation of one line is off.

eric-wieser replied:
How were you able to resolve the warning, @pdebuyl?

pdebuyl replied:
By indenting the comment marker by two extra spaces.

pdebuyl committed:
remove blank line after 'for doctests'

pdebuyl committed:
add one space to indent below "for doctests"

pdebuyl commented:
I can only get the "original" version to pass:

```
.. for doctests

   >>> import numpy.distutils.misc_util
   >>> config = np.distutils.misc_util.Configuration(None, '', '.')

  >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```

Should I revert to it?

charris commented on doc/source/reference/distutils.rst:
```diff
@@ -117,6 +120,10 @@ install the C library, you just use the method `add_installed_library` instead o
+    >>> import numpy.distutils.misc_util
+    >>> config = np.distutils.misc_util.Configuration(None, '', '.')
+
   >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```
This line needs to be correctly indented.

pdebuyl committed:
modify indent of 'for doctests' code

pdebuyl commented:
For info, the travis build succeeded (even though the CI badge is not updated on this PR page, you can simply click on it).

Also, I can squash the PR when/if necessary.

The current indent is

```
  .. for doctests
    >>> import numpy.distutils.misc_util
    >>> config = np.distutils.misc_util.Configuration(None, '', '.')

  >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```

(I indented the whole `..` block)

charris commented:
OK, that works. But look where `.. for doctests` is located in

- `doc/source/reference/arrays.classes.rst`
- `doc/source/reference/maskedarray.baseclass.rst`

pdebuyl commented:
Initially, the `for doctests` was at the top. I moved it on request https://github.com/numpy/numpy/pull/15616#pullrequestreview-361864421

I'll let you (i.e. NumPy maintainers) decide where it should ultimately go.

charris commented:
@eric-wieser Ping.

pdebuyl commented:
Hi @eric-wieser @charris 

The choice is between "import in comment at the top of the file" or "import in comment above line that requires it". I have no preference but this seems like a minor issue that could get merged without requiring too much of maintainer time, decreasing the PR count by 1 :-)

eric-wieser commented on doc/source/reference/distutils.rst:
```diff
@@ -1,3 +1,6 @@
+.. for doctests
+   >>> with open('foo.c', 'w') as f: pass
+
```
Seems odd to leave this at the top while the others are at the bottom

pdebuyl replied:
Fair enough, it will move :-)

pdebuyl replied:
It is even more useful that the import and the creation of the file foo.c are necessary at the same place.

eric-wieser commented:
@charris: Putting the hidden doctest code right before it's needed makes it easier to move doc content around or delete it without leaving behind stale code at the top the file - so I prefer the approach taken by this PR.

eric-wieser commented on doc/source/reference/distutils.rst:
```diff
@@ -117,6 +120,10 @@ install the C library, you just use the method `add_installed_library` instead o
 `add_library`, which takes the same arguments except for an additional
 ``install_dir`` argument::
 
+  .. for doctests
```
Perhaps we should be more explicit here:
```suggestion
  .. hidden in a comment so as to be included in refguide but not rendered documentation
```

pdebuyl committed:
Update doc/source/reference/distutils.rst

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

pdebuyl committed:
move creation of foo.c

eric-wieser reviewed (approved):
Looks good - can you check that this passes refguide still, assuming that CI does not yet check that?

pdebuyl commented:
It passes the `refguide_check.py --rst` with current master. Should I squash now?

mattip commented:
Thanks @pdebuyl, github can do the squash-merge.

pdebuyl commented:
Thanks :-)

eric-wieser commented:
@mattip: is the some way we can maintain a blacklist of broken rst files in git, and remove them one at a time alongside these PRs?
